lib/spack/spack/build_systems/perl.py

```diff
@@ -0,0 +1,117 @@
+
+import inspect
+
+from spack.directives import extends
+from spack.package import PackageBase, run_after, InstallError
+from spack.util.executable import Executable
+
+
+class PerlPackage(PackageBase):
+    """Specialized class for packages that are built using Perl
+
+    This class provides three phases that can be overridden:
+
+        1. :py:meth:`~.PerlPackage.configure`
+        2. :py:meth:`~.PerlPackage.build`
+        3. :py:meth:`~.PerlPackage.install`
+
+    They all have sensible defaults. Some packages may also need to override:
+
+        +-----------------------------------------+--------------------------+
+        | **Method**                              | **Purpose**              |
+        +=========================================+==========================+
+        | :py:attr:`~.PerlPackage.build_method    | Perl build method:       |
+        |                                         | 'Makefile.PL' (default)  |
+        |                                         | or 'Build.PL'            |
+        +-----------------------------------------+--------------------------+
+        | :py:meth:`~.PerlPackage.configure_args` | List of arguments for    |
+        |                                         | Makefile.PL or Build.PL  |
+        |                                         | (excluding prefix)       |
+        +-----------------------------------------+--------------------------+
+    """
+    #: Phases of a Perl package
+    phases = ['configure', 'build', 'install']
+
+    #: This attribute is used in UI queries that need to know the build
+    #: system base class
+    build_system_class = 'PerlPackage'
+
+    #: This attribute is used to select the perl build method.
+    #: Supported types are 'Makefile.PL' and 'Build.PL'.
+    build_method = 'Makefile.PL'
+
+    extends('perl')
+
+    def configure_args(self):
+        """Produces a list containing the arguments that must be passed to
+        the configure method. The installation prefix is always prepended.
+
+        :return: list of arguments for Makefile.PL or Build.PL
+        """
+        return []
+
+    def configure(self, spec, prefix):
+        """Runs Makefile.PL or Build.PL with the arguments specified in
+        :py:meth:`.configure_args` and an appropriate installation prefix.
+        """
+        if self.build_method == 'Makefile.PL':
+            options = ['Makefile.PL', 'INSTALL_BASE={0}'.format(prefix)]
+        elif self.build_method == 'Build.PL':
+            options = ['Build.PL', '--install_base', prefix]
+        else:
+            raise InstallError('Unknown build_method for perl package')
+        options += self.configure_args()
+        inspect.getmodule(self).perl(*options)
+
+    def build(self, spec, prefix):
+        """Builds a Perl package."""
+        if self.build_method == 'Makefile.PL':
+            makex = inspect.getmodule(self).make
+            makex()
+            if self.run_tests:
+                makex('test')
```

Comment:
Let's move the `test` step to a separate function. See `AutotoolsPackage` for how this is done. The benefit of this is that packages that don't provide a `test` target can skip this. Also, if there are any additional/different tests to run, users can override this by adding a `test(self)` function to their package.